pkg/asset/machines/openstack/openstackmachines.go

```diff
@@ -0,0 +1,215 @@
+	var ports []capo.PortOpts
+
+	if platform.ControlPlanePort != nil {
+		networkID := platform.ControlPlanePort.Network.ID
+
+		// In a dual-stack cluster, when network ID or Name is not specified, the network ID needs to
+		// be discovered and added to the ProviderSpec for CAPO to create one unique Port with two addresses.
+		var err error
+		if networkID == "" && platform.ControlPlanePort.Network.Name == "" && len(platform.ControlPlanePort.FixedIPs) == 2 {
+			networkID, err = getNetworkFromSubnet(platform.ControlPlanePort.FixedIPs[0], platform.Cloud)
+			if err != nil {
+				return nil, err
+			}
+		}
+
+		for _, fixedIP := range platform.ControlPlanePort.FixedIPs {
+			port := capo.PortOpts{
+				FixedIPs: []capo.FixedIP{
+					{
+						Subnet: &capo.SubnetFilter{
+							ID:   fixedIP.Subnet.ID,
+							Name: fixedIP.Subnet.Name,
+						},
+					},
+				},
+				Network: &capo.NetworkFilter{
+					Name: platform.ControlPlanePort.Network.Name,
+					ID:   networkID,
+				},
+			}
+			ports = append(ports, port)
+		}
+	} else {
+		port := capo.PortOpts{
+			FixedIPs: []capo.FixedIP{
+				{
+					Subnet: &capo.SubnetFilter{
+						Name: fmt.Sprintf("%s-nodes", clusterID),
+						Tags: fmt.Sprintf("openshiftClusterID=%s", clusterID),
+					},
+				},
+			},
+		}
+		ports = append(ports, port)
+	}
+
+	for _, networkID := range mpool.AdditionalNetworkIDs {
+		ports = append(ports, capo.PortOpts{
+			Network: &capo.NetworkFilter{
+				ID: networkID,
+			},
+		})
+	}
+
+	securityGroups := []capo.SecurityGroupFilter{
+		{
+			Name: fmt.Sprintf("%s-%s", clusterID, role),
+		},
+	}
+
+	for _, securityGroup := range mpool.AdditionalSecurityGroupIDs {
+		securityGroups = append(securityGroups, capo.SecurityGroupFilter{ID: securityGroup})
+	}
+
+	// FIXME: Uncomment when the server group rework merged
+	// https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/1779
+	// serverGroupName := clusterID + "-" + role
+	// We initially used the AZ name as part of the server group name for the masters
+	// but we realized that it was not useful. Whether or not the AZ is specified, the
+	// masters will be spread across multiple hosts by default by the Nova scheduler
+	// (the policy can be changed via `serverGroupPolicy` in install-config.yaml).
+	// For the workers, we still use the AZ name as part of the server group name
+	// so the user can control the scheduling policy per AZ and change the MachineSets
+	// if needed on a day 2 operation.
+	// FIXME: Uncomment when we pass failure domain info
+	//if role == "worker" && failureDomain.AvailabilityZone != "" {
+	//	serverGroupName += "-" + failureDomain.AvailabilityZone
+	//}
+
+	spec := capo.OpenStackMachineSpec{
+		// FIXME: Add failure domains/AZ support
+		CloudName: CloudName,
+		Flavor:    mpool.FlavorName,
+		// TODO: Are we creating the credentials request in the openshift-cluster-api namespace?
+		IdentityRef: &capo.OpenStackIdentityReference{
```

Comment:
Do we need to somehow specify `UserDataSecret` as well?
https://github.com/openshift/installer/blob/master/pkg/asset/machines/openstack/machines.go#L252